geode-core/src/test/java/org/apache/geode/internal/cache/execute/PartitionedRegionFunctionResultSenderTest.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.geode.internal.cache.execute;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+
+import java.util.concurrent.TimeUnit;
+
+import org.junit.jupiter.api.Test;
+
+import org.apache.geode.cache.execute.Function;
+import org.apache.geode.cache.execute.FunctionContext;
+import org.apache.geode.cache.execute.FunctionException;
+import org.apache.geode.cache.execute.ResultCollector;
+import org.apache.geode.distributed.DistributedMember;
+import org.apache.geode.distributed.internal.DistributionManager;
+import org.apache.geode.internal.cache.PartitionedRegion;
+import org.apache.geode.internal.cache.execute.metrics.FunctionStats;
+
+public class PartitionedRegionFunctionResultSenderTest {
+  DistributionManager dm = mock(DistributionManager.class);
+  PartitionedRegion region = mock(PartitionedRegion.class);
+  ATestResultCollector rc = new ATestResultCollector();
+  ServerToClientFunctionResultSender serverToClientFunctionResultSender =
+      mock(ServerToClientFunctionResultSender.class);
+  FunctionStats functionStats = mock(FunctionStats.class);
+
+  @Test
+  public void whenResponseToClientInLastResultFailsEndResultsIsCalled_OnlyLocal_NotOnlyRemote() {
+    doThrow(new FunctionException()).when(serverToClientFunctionResultSender)
+        .lastResult(any(), any());
+    PartitionedRegionFunctionResultSender sender =
+        new PartitionedRegionFunctionResultSender(dm,
+            region, 1, rc, serverToClientFunctionResultSender, true, false, true,
+            new TestFunction(), new int[2]);
+
+    try {
+      sender.lastResult(new FunctionException());
+    } catch (FunctionException expected) {
+    }
```

Comment:
You should always add assertions for expected exceptions:
```
Throwable thrown = catchThrowable(() -> {
  sender.lastResult(new FunctionException());
}

assertThat(thrown)
  .isInstanceOf(FunctionException.class)
  .(any other assertions that are valuable?)
```

Reply:
I did not add the assertion because it was not something I wanted to verify in the test. Nevertheless, thanks to your comment I have seen that the exception is not thrown so I have removed it.
I have also changed the test cases so that instead of sending an exception, an object is sent.